File: 0875-koko-eating-bananas/0875-koko-eating-bananas.ts
function minEatingSpeed(piles: number[], h: number): number {
        // left is equal to 1 (because need to eat at least 1 banana)
        let left: number = 1
        // right is equal to max(piles) - last index
        let right: number = Math.max(...piles);
        // res = right
        let res: number = right;

        // while left is less than or equal to right:
            // mid = left + right / 2
            // hours = 0
            // go to every pile:
            //     hours += Math.ceil(pile / mid)
            // if hours <= h:
            //      update res
            //      right = mid - 1 (search left portion)
            // else:
            //      left = mid + 1 (search for right portion)
        while (left <= right) {
            let mid: number = Math.floor((left + right) /2);
            let hours: number = 0;
            for (const pile of piles) {
                hours += Math.ceil(pile / mid);
            }
            if (hours <= h) {
                res = Math.min(res,mid);
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }      

        // return res
        return res;
};

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
class Solution:
    # Time Complexity: O(n * log M) where n is the number of piles, and M is the maximum pile size.
    # Space Complexity: O(1) (constant space).
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        left = 1
        right = max(piles)
        res = right
        
        while left <= right:
            mid = (left + right) // 2
            hours = 0
            for pile in piles:
                hours += math.ceil(pile/mid)
            if hours <= h:
                res = mid
                right = mid - 1 
            else:
                left = mid + 1
        return res
        
        
        
        
        
        
        
        
        
#         left = 1
#         right = max(piles)
#         res = right
        
#         while left <= right:
#             mid = (left + right) // 2
#             hours = 0
#             for pile in piles:
#                 hours += math.ceil(pile/mid)
#             if hours <= h:
#                 res = mid
#                 right = mid - 1
#             else:
#                 left = mid + 1
        
#         return res
        
